### PR TITLE
Add regexp operator to format filters

### PR DESCRIPTION
**Pull request by lukasfink1**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
This adds a new format filter operater that allows to filter video formats by matching the string based format fields against a regex.

Example usage: -f `bestvideo*[vcodec~="(he|a)vc|[hx]26[45]"]+bestaudio/best`

Alternatively, single quotes can be used to delimit the regex.

This implements my feature request #2681.

### CODE REVIEW

**Comment by pukkandan** [comment]
We can allow quoting for the other operators as well. This should simplify the code too

**Comment by lukasfink1** [comment]
Yes, this seems like a good idea. I’ll look into that.

File: yt_dlp/YoutubeDL.py [hunk 1]
```diff
@@ -1819,8 +1819,8 @@ def _build_format_filter(self, filter_spec):
         }
         operator_rex = re.compile(r'''(?x)\s*
             (?P<key>width|height|tbr|abr|vbr|asr|filesize|filesize_approx|fps)\s*
-            (?P<op>%s)(?P<none_inclusive>\s*\?)?\s*
-            (?P<value>[0-9.]+(?:[kKmMgGtTpPeEzZyY]i?[Bb]?)?)\s*
+            (?P<op>%s)\s* (?P<none_inclusive>\?\s*)?
+            (?P<value> [0-9.]+ (?:[kKmMgGtTpPeEzZyY]i?[Bb]?)?)\s*
```
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:1823]: unnecessary diff

File: yt_dlp/YoutubeDL.py [hunk 2]
```diff
@@ -1842,15 +1842,21 @@ def _build_format_filter(self, filter_spec):
...
+                '~=': lambda attr, value: value.search(attr) is not None
             }
             str_operator_rex = re.compile(r'''(?x)\s*
                 (?P<key>[a-zA-Z0-9._-]+)\s*
-                (?P<negation>!\s*)?(?P<op>%s)(?P<none_inclusive>\s*\?)?\s*
-                (?P<value>[a-zA-Z0-9._-]+)\s*
+                (?P<negation>!\s*)? (?P<op>%s)\s* (?P<none_inclusive>\?\s*)?
+                (?P<quote>["'])?
+                (?P<value> (?(quote) (?:(?!(?P=quote))[^\\]|\\.)+ | [a-zA-Z0-9._-]+ ))
+                (?(quote) (?P=quote) )\s*
```
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:1852]:
```suggestion
                (?P<negation>!\s*)?(?P<op>%s)\s*(?P<none_inclusive>\?\s*)?
                (?P<quote>["'])?
                (?P<value>(?(quote)(?:(?!(?P=quote))[^\\]|\\.)+|[\w.-]+))
                (?(quote)(?P=quote))\s*
```

File: yt_dlp/YoutubeDL.py [hunk 3]
```diff
@@ -1842,15 +1842,21 @@ def _build_format_filter(self, filter_spec):
...
                 ''' % '|'.join(map(re.escape, STR_OPERATORS.keys())))
             m = str_operator_rex.fullmatch(filter_spec)
             if m:
-                comparison_value = m.group('value')
+                if m.group('op') == '~=':
+                    comparison_value = re.compile(m.group('value'))
+                else:
+                    comparison_value = m.group('value')
```
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:1859]: You have to remove the escape chars before this

**Comment by pukkandan** [comment]
Please avoid force-pushing during review
